### PR TITLE
Allocation autocomplete, client api

### PR DESCRIPTION
**Pull request by chelseakomlo**
This PR introduces two changes: 
1. Autocomplete of `nomad alloc-status` for allocation ids
2. API pkg for resource list

One thing to consider is that we use the term `resource` in #2964, but this term already exists in for machine resources in `/api/resources`. I used `/api/job_resources` to differentiate, but let me know if this should be named something differently or included in `/api/resources`

### CODE REVIEW

File: api/job_resources.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package api
+
+import (
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type JobResources struct {
```
Comment (dadgar) [inline api/job_resources.go]: Why `JobResources` and not just `Resources`. The object can list more than just jobs.

File: api/job_resources.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	client *Client
+}
+
+// JobResources returns a handle on the JobResources endpoints
+func (c *Client) JobResources() *JobResources {
+	return &JobResources{client: c}
+}
+
+// List returns a list of all resources for a particular context. If a
+// context is not specified, matches for all contezts are returned.
```
Comment (dadgar) [inline api/job_resources.go]: contezts -> contexts

File: api/job_resources.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+func (j *JobResources) List(prefix string, context string) (*structs.ResourceListResponse, error) {
```
Comment (dadgar) [inline api/job_resources.go]: Don't want to take or return anything from `structs` in this package. Create a copy of the ListResponse here.

File: nomad/resources_endpoint.go [hunk 4]
```diff
@@ -73,6 +73,15 @@ func getResourceIter(context, prefix string, ws memdb.WatchSet, state *state.Sta
 	}
 }
 
+// If the length of a string is odd, return a subset of the string to the last
+// even character
+func roundDownIfOdd(s string) string {
```
Comment (dadgar) [inline nomad/resources_endpoint.go]: I am fine with this if it is a place holder to a slightly more advanced implementation. My concern here is two fold:

1) Jobs don't need to be rounded, only UUIDs.
2) If I do a prefix of ab1 and I get a return of: ["ab12..", "ab23...", "ab34", ...] something seems broken :)

File: api/job_resources.go (hunk 1 above)
Comment (chelseakomlo) [inline api/job_resources.go]: Yes, that was one thing I wasn't sure about, as currently a resource in the api package is commented as "Resource encapsulates the required resources of a given task or task group."

File: nomad/resources_endpoint.go (hunk 4 above)
Comment (chelseakomlo) [inline nomad/resources_endpoint.go]: Ok, understood. So we can round the UUID for the query, and then use a filter function for the remaining character.

File: api/job_resources.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	return &JobResources{client: c}
+}
+
+// List returns a list of all resources for a particular context. If a
+// context is not specified, matches for all contezts are returned.
+func (j *JobResources) List(prefix, context string) (*structs.ResourceListResponse, error) {
+	var resp structs.ResourceListResponse
+	req := &structs.ResourceListRequest{Prefix: prefix, Context: context}
+
+	_, err := j.client.write("/v1/resources/", req, &resp, nil)
```
Comment (jippi) [inline api/job_resources.go]: is `/v1/resources` a new endpoint?

Comment (chelseakomlo) [inline api/job_resources.go]: We renamed it to /v1/cluster/search in a later commit, but this is a new endpoint.

File: api/search.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+// Search returns a handle on the Search endpoints
+func (c *Client) Search() *Search {
+	return &Search{client: c}
+}
+
+// List returns a list of matches for a particular context and prefix. If a
+// context is not specified, matches for all contexts are returned.
+func (s *Search) List(prefix, context string) (*structs.SearchResponse, error) {
```
Comment (dadgar) [inline api/search.go]: List -> PrefixSearch

File: api/search.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package api
+
+import (
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
```
Comment (dadgar) [inline api/search.go]: Can we add consts for the various contexts and use them in the commands pkg

File: nomad/search_endpoint.go [hunk 8]
```diff
@@ -73,10 +82,23 @@ func getResourceIter(context, prefix string, ws memdb.WatchSet, state *state.Sta
...
+// This only applies to UUIDs, Job names are excluded
+func roundUUIDDownIfOdd(prefix, context string) string {
+	if context == "job" {
+		return prefix
+	}
+
+	if len(prefix)%2 == 0 {
+		return prefix
+	}
+	return prefix[:len(prefix)-1]
```
Comment (dadgar) [inline nomad/search_endpoint.go]:
```
l := len(prefix)
return prefix[:l-l%2]
```

File: nomad/search_endpoint.go [hunk 9]
```diff
@@ -46,10 +51,14 @@ func (r *Resources) getMatches(iter memdb.ResultIterator) ([]string, bool) {
 		case *structs.Node:
 			id = raw.(*structs.Node).ID
 		default:
-			r.srv.logger.Printf("[ERR] nomad.resources: unexpected type for resources context: %T", t)
+			s.srv.logger.Printf("[ERR] nomad.resources: unexpected type for resources context: %T", t)
 			continue
 		}
 
+		if !isSubset(prefix, id) {
+			break
```
Comment (dadgar) [inline nomad/search_endpoint.go]: break or continue?

Comment (chelseakomlo) [inline nomad/search_endpoint.go]: Yes, I caught this after pushing.

File: api/search.go (hunk 7 above)
Comment (chelseakomlo) [inline api/search.go]: Definitely. These should live in the api package though?

Comment (dadgar) [inline api/search.go]: Probably both where API pkg defines them from structs

File: nomad/search_endpoint.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/nomad/nomad/state"
+	st "github.com/hashicorp/nomad/nomad/structs"
+)
+
+// truncateLimit is the maximum number of matches that will be returned for a
```
Comment (dadgar) [inline nomad/search_endpoint.go]: Put comment over field

Comment (dadgar) [inline nomad/search_endpoint.go]: And below

File: nomad/search_endpoint.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
+package nomad
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/nomad/nomad/state"
+	st "github.com/hashicorp/nomad/nomad/structs"
```
Comment (dadgar) [inline nomad/search_endpoint.go]: Can you keep it as structs. Avoid named imports unless there is a conflict or the name is terribly long

File: nomad/search_endpoint.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+		case *st.Allocation:
+			id = raw.(*st.Allocation).ID
+		case *st.Node:
+			id = raw.(*st.Node).ID
+		default:
+			s.srv.logger.Printf("[ERR] nomad.resources: unexpected type for resources context: %T", t)
+			continue
+		}
+
+		if !isSubset(prefix, id) {
```
Comment (dadgar) [inline nomad/search_endpoint.go]: https://golang.org/pkg/strings/#HasPrefix

File: nomad/search_endpoint.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+		run: func(ws memdb.WatchSet, state *state.StateStore) error {
+
+			iters := make(map[st.Context]memdb.ResultIterator)
+
+			contexts := allContexts
+			if args.Context != st.All {
+				contexts = []st.Context{args.Context}
+			}
+
+			for _, e := range contexts {
```
Comment (dadgar) [inline nomad/search_endpoint.go]: `e` -> `context` or `ctx`

File: nomad/structs/structs.go [hunk 14]
```diff
@@ -89,6 +89,17 @@ const (
 	GetterModeDir  = "dir"
 )
 
+// Context is a type which is searchable via a unique identifier.
```
Comment (dadgar) [inline nomad/structs/structs.go]: Comment doesn't make much sense.

`Context defines the scope in which a search for Nomad object operates`

File: api/search.go [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	client *Client
+}
+
+// Search returns a handle on the Search endpoints
+func (c *Client) Search() *Search {
+	return &Search{client: c}
+}
+
+// PrefixSearch returns a list of matches for a particular context and prefix. If a
+// context is not specified, matches for all contexts are returned.
```
Comment (dadgar) [inline api/search.go]: Nit: Might be worth deleting the last line of the comment since we now assume there will always be a context, just `contexts.All`

**Pull request merged**